### ISSUE TITLE
Incredibly slow startup

### ISSUE THREAD

**Issue by AhmedAA** [syl20bnr/spacemacs#8144]
Spacemacs takes about 1 minute (stable) to start up.
This renders it virtually unusable..
I tried adding the helm fix to my .spacemacs, but it did not help.
I also tried adding the lines to my config in .ssh, which also did not help.

I know the problem lies in auto-layer.el, as spacemacs completely hangs when it loads from it, but I got no clue why..

My .spacemacs can be found here https://github.com/AhmedAA/configz/blob/master/dotfiles/emacs/dotspacemacs

I usually use the develop branch, but switched to master to see if the issue is still present, and it is.
I use Emacs 25.1.1
I tried using vanilla Emacs (i.e. without any spacemacs checked out) and it started instantly.

I am kind of stuck now, as I have no idea what to do to fix this issue, and it is really annoying :S

Any help appreciated!

**Comment by bmag** [comment]
Sorry to hear that. If startup time is a problem, consider running an instance of `emacs --daemon` in the background and connecting to it with `emacsclient`.
As for solving the startup time, how do you know the problem lies in `auto-layer.el`? You can set `dotspacemacs-verbose-loading` to `t` (there's already an entry in your dotspacemacs, you need to change it) and run `emacs --timed-requires`. Spacemacs should show you how much loading each part took. If the info doesn't appear in the home buffer, look for it in the messages buffer (`*Message*`).

**Comment by AhmedAA** [comment]
Perhaps I should have said that I am 'fairly certain the problem lies in auto-layer', and not that I know it does, sorry about the wording.
I tried running with ```emacs --daemon``` but it simply makes no sense in my usecase (and my system started hanging on poweroff if I didn't kill the daemon before shutdown, which is a different problem I didn't want to debug).

Anyway, the results from timed-requires:
```Threshold set at 0.150 seconds

File /home/ahmed/.emacs.d/init.el: Required spacemacs-purpose-popwin: 0.199 sec
File /home/ahmed/.emacs.d/elpa/helm-20170115.1348/helm-net.elc: Required url: 0.181 sec
File /home/ahmed/.emacs.d/elpa/helm-20170115.1348/helm-external.elc: Required helm-net: 0.206 sec
File /home/ahmed/.emacs.d/elpa/helm-20170115.1348/helm-files.elc: Required helm-external: 0.212 sec
File /home/ahmed/.emacs.d/elpa/helm-20170115.1348/helm-mode.elc: Required helm-files: 0.468 sec
File nil: Required helm: 0.540 sec```

which in my opinion makes no sense, as this doesn't add up to ~60 seconds..
And from the message buffer:
```Loading /home/ahmed/.emacs.d/core/core-load-paths.el (source)...done
Loading /home/ahmed/.spacemacs...done
Setting the font...
PNG warning: Interlace handling should be turned on when using png_read_image
Open the quickhelp.
Loading /home/ahmed/.emacs.d/layers/+themes/themes-megapack/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+completion/auto-completion/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+tools/docker/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/emacs-lisp/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+source-control/git/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+source-control/github/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/go/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+completion/helm/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/html/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+emacs/org/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+checkers/syntax-checking/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+source-control/version-control/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/latex/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/javascript/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/python/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+distributions/spacemacs/layers.el (source)...
Loading /home/ahmed/.emacs.d/layers/+distributions/spacemacs-base/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-completion/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-layouts/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-editing/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-editing-visual/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-evil/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-language/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-misc/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-purpose/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-ui/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-ui-visual/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-org/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+distributions/spacemacs/layers.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+distributions/spacemacs-bootstrap/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+distributions/spacemacs-bootstrap/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-ui-visual/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-purpose/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-evil/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-editing/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-layouts/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-completion/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+distributions/spacemacs-base/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+completion/auto-completion/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/emacs-lisp/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+source-control/git/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/go/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+completion/helm/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/html/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+emacs/org/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+checkers/syntax-checking/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+source-control/version-control/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/latex/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/javascript/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/python/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+distributions/spacemacs-bootstrap/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-ui-visual/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-layouts/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-completion/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+distributions/spacemacs-base/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+completion/auto-completion/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/emacs-lisp/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+source-control/git/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/go/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+emacs/org/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+checkers/syntax-checking/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+source-control/version-control/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/latex/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/javascript/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/python/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/auto-layer.el (source)...done
.emacs.d/layers/+spacemacs/spacemacs-purpose/local/spacemacs-purpose-popwin/spacemacs-purpose-popwin.el: Obsolete name arg "pupo" to constructor purpose-conf
Loading /home/ahmed/.emacs.d/layers/+distributions/spacemacs-base/keybindings.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+source-control/version-control/keybindings.el (source)...done
Loading /home/ahmed/.emacs.d/.cache/spacemacs-buffer.el (source)...done
Spacemacs is ready.
Skipping check for new version (reason: dotfile)
Loading /home/ahmed/.emacs.d/.cache/recentf...done
Open the quickhelp.```

I could try running vanilla spacemacs, with no modifications and see if I have the same issues? I am pretty sure I am running on decent dotfile, but there might be something I missed?

And thanks for helping :)

**Comment by bmag** [comment]
> I could try running vanilla spacemacs, with no modifications and see if I have the same issues?

Yes, please do so.

> I could try running vanilla spacemacs, with no modifications and see if I have the same issues?

Have a look in `FAQ.org` (also available on spacemacs.org), in case something there is relevant.

> which in my opinion makes no sense, as this doesn't add up to ~60 seconds..

So (maybe) the delay is caused (seemingly) immediately after startup. Helm, for example, is loaded ~1 second after startup has finished, and it has caused delays for people in the past. There's a FAQ entry about Helm slowness on startup.

Can you also post the output of `SPC h d s` (shows your system info)

**Comment by AhmedAA** [comment]
Vanilla spacemacs (30 sec startup).
Really flustered about this..

#### System Info :computer:
- OS: gnu/linux
- Emacs: 25.1.1
- Spacemacs: 0.200.7
- Spacemacs branch: master (rev. c2774bc0)
- Graphic display: t
- Distribution: spacemacs
- Editing style: emacs
- Completion: helm
- Layers:
```elisp
(helm emacs-lisp)
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG IMAGEMAGICK SOUND GPM DBUS GCONF GSETTINGS NOTIFY ACL GNUTLS LIBXML2 FREETYPE M17N_FLT LIBOTF XFT ZLIB TOOLKIT_SCROLL_BARS GTK3 X11

Load times:
```
Threshold set at 0.150 seconds

File /home/ahmed/.emacs.d/elpa/helm-20170116.2331/helm-mode.elc: Required helm-files: 0.238 sec
File nil: Required helm: 0.283 sec
```

I already tried the Helm fixes from the FAQ, and I really couldn't find other entries in the FAQ regarding slow startup other than the Helm issue :S
Specifically, I added the lisp snippet under ```user-init``` and I also tried under ```user-config```. Nothing changed the startup time. The only thing that did change it, was if I cut in the number of layers, but the reduction wasn't really noteworthy..

**Comment by bmag** [comment]
I'm not sure how we can solve it. We can start with print-debugging to find where the bottleneck is. By adding this:
```elisp
(message "In <some place>: %s" (current-time-string))
```
to different places, you can look in the `*Messages*` buffer to see what part took the most time. Interesting places to put this are the different fucntions in ~.spacemacs~ (or ~/.spacemacs.d/init.el`), and between expressions in `.emacs.d/init.el`.

**Comment by deb0ch** [comment]
One or two years ago, in vanilla Emacs, I and some friends used to have a problem with Emacs taking sometimes 60s+ to startup because of some network issues that we never dug into.

It just used to happen depending on the state of the network: it never happened when unconnected, sometimes when connected and often if connected without access to Internet.

This is very vague but maybe it could be related (?)

**Comment by AhmedAA** [comment]
> One or two years ago, in vanilla Emacs, I and some friends used to have a problem with Emacs taking sometimes 60s+ to startup because of some network issues that we never dug into.
> 
> It just used to happen depending on the state of the network: it never happened when unconnected, sometimes when connected and often if connected without internet access.
> 
> This is very vague but maybe it could be related (?)

Just tried, had no effect. Interesting bug though..
I think I am just incredibly confused, as the only thing I can point at right now, is my box's setup. If something is wrongly configured or anything, but I have never experienced this.

**Comment by AhmedAA** [comment]
Now this is interesting..
Seems like it takes quite a while to load layers:
```
In spacemacs init: Fri Jan 20 12:21:01 2017
In dotfile check: Fri Jan 20 12:21:01 2017
In layers: Fri Jan 20 12:21:01 2017
Loading /home/ahmed/.emacs.d/layers/+completion/helm/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/emacs-lisp/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+distributions/spacemacs/layers.el (source)...
Loading /home/ahmed/.emacs.d/layers/+distributions/spacemacs-base/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-completion/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-layouts/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-editing/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-editing-visual/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-evil/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-language/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-misc/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-ui/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-ui-visual/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-org/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+distributions/spacemacs/layers.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+distributions/spacemacs-bootstrap/packages.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+distributions/spacemacs-bootstrap/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-ui-visual/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-evil/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-editing/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-layouts/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-completion/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+distributions/spacemacs-base/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+completion/helm/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/emacs-lisp/funcs.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+distributions/spacemacs-bootstrap/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-ui-visual/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-layouts/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+spacemacs/spacemacs-completion/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+distributions/spacemacs-base/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+lang/emacs-lisp/config.el (source)...done
Loading /home/ahmed/.emacs.d/layers/auto-layer.el (source)...done
Loading /home/ahmed/.emacs.d/layers/+distributions/spacemacs-base/keybindings.el (source)...done
In sync layers: Fri Jan 20 12:21:32 2017
Loading /home/ahmed/.emacs.d/.cache/spacemacs-buffer.el (source)...done
In buffer info box: Fri Jan 20 12:21:32 2017
In server startup hook: Fri Jan 20 12:21:32 2017
In require server: Fri Jan 20 12:21:32 2017
In server start: Fri Jan 20 12:21:32 2017
Spacemacs is ready.
In user-config: Fri Jan 20 12:21:32 2017
```
Never had this issue before though, any ideas?

**Comment by AhmedAA** [comment]
Arch Linux just bumped emacs to version 25.1.1, and I still have this issue (with a fresh spacemacs install).

**Comment by syl20bnr** [comment]
@AhmedAA sorry for the late reply, we have some useful command line arguments you ca try to debug.
First when you launch emacs with `--debug-init` the configuration layer system switch to verbose mode and you get more information in the message buffer.
I just see that we have not documented any of the debug command line parameter (!!), my bad, it should not happen, going to fix this right now, I'll post back the link to the file when I'm done.

**Comment by AhmedAA** [comment]
@syl20bnr first of all, thank you very much for spacemacs :) can't even begin to describe how grateful I am for this project.

It seems I get a lot of information this way, and I have to admit that I have a hard time deciphering the relevant parts :S should I create a gist or something?

Thanks for your time!

**Comment by AhmedAA** [comment]
Btw, I am pretty sure the issue is Arch Linux specific.
I haven't experienced this issue on other distros, but I will test it out during the weekend.
I got no clue what Arch Linux does though..

**Comment by zhzhxtrrk** [comment]
I'm using Spacemacs on Windows 10, it takes sooooo long time(about 1 minute) to startup too.

**Comment by yenda** [comment]
Have you tried excluding the exec-path-from-shell package ?
It seems that you are also victim of this mac-only package that is added by default and ruins startup on other systems.

https://github.com/syl20bnr/spacemacs/issues/7588

**Comment by AhmedAA** [comment]
@yenda 
Thank you so much!
This fixed the issue.

However, a side note:
If a layer uses the package, that layer breaks.. for instance, I cannot use the golang layer now, which sucks.. so I need to notify the maintainer of that layer to make some kind of fallback.

TL;DR (for future readers):
exclude `exec-path-from-shell` in your .spacemacs, under excluded packages
check if you have a layer enabled that uses that package, and remove that layer
???
enjoy fast emacs again!

**Comment by aalireza** [comment]
@yenda Excluding `exec-path-from-shell` causes at least one problem if you have the `go` layer since it tries to execute `exec-path-from-shell-copy-env`. 

For some reason though if you don't use gui and enter `emacs24 -nw` in the terminal, the startup time would still be slow. It consistently takes around 13 seconds to load all the packages `301 packages loaded in 130.889s  (e:236 r:2 l:14 b:49)`, as opposed to around 5 seconds which is what it used to be before I reinstalled Ubuntu.
